Better handling of `opam show` errors

hi, 

i just ran into problems with odoc/odig, and it turned out the cause was that `opam show` does not handle empty `.opam` files gracefully.  this is described here: https://github.com/b0-system/odig/issues/51

the way i got an empty opam file was by calling `dune install` in a small local project of mine which defines one executable, and had an empty opam file in the source directory.

is this a dune problem or an opam problem or is it as intended?

```
# opam config report
# opam-version      2.0.4
# self-upgrade      no
# system            arch=x86_64 os=macos os-distribution=homebrew os-version=10.14.5
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              4
# repositories      1 (http) (default repo at 24059019)
# pinned            6 (git)
# current-switch    4.07.1+flambda
```

> i just ran into problems with odoc/odig, and it turned out the cause was that `opam show` does not handle empty `.opam` files gracefully. this is described here: [b0-system/odig#51](https://github.com/b0-system/odig/issues/51)

Just to make things a bit clearer here. `odig` is invoking `opam show` on the raw `opam` files installed in the library directory as [per convention](https://b0-system.github.io/odig/doc/odig/packaging.html#meta) to get metadata for installed packages. 

Since `opam show` does have some overhead (~200ms on opam 2.0.4) it does not do so individually on each `opam` file but on all the files at the same time ([here](https://github.com/b0-system/odig/blob/b9add704519889079dfb633d6ac1f5d1c9840683/src/odig_support.ml#L237-L267)) using the equivalent of:
```
'opam' 'show' '--normalise' '--no-lint' \
'--field=name:,authors:,bug-reports:,depends:,devrepo:,doc:,homepage:,license:,maintainer:,synopsis:,tags:,version:' \
$(LIBDIR)/*/opam
```
 `odig` is robust in case this invocation fails. However if a single of the files has a parse error (e.g. an empty `opam` file) then all the metadata for the others is lost because the tool just exits with 99 and for example:
```
[ERROR] At /private/tmp/opam:2:3:
        Parse error [skipped]

Fatal error:
At /private/tmp/opam:2:3:
Backtrace:
  Raised at file "src/format/opamPp.ml", line 76, characters 45-66
  Called from file "src/format/opamFile.ml", line 874, characters 45-76
  Called from file "src/format/opamFile.ml", line 130, characters 16-33
[...]
```
I think it would be nicer to output the data for opam files that do parse, to report those that do not with an error on stderr and eventually exit with 30 that did happen.

(N.B. in case you wonder why odig doesn't use the `opam` API, it's to avoid tight coupling of the tools -- that is `odig` can work even if `opam` is not there).

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -703,10 +703,7 @@ let show =
   let sort = mk_flag ["sort"] "Sort opam fields" in
   let opam_files_in_dir d =
     match OpamPinned.files_in_source d with
-    | [] ->
-      OpamConsole.warning "No opam files found in %s"
-        (OpamFilename.Dir.to_string d);
-      []
+    | [] -> []
     | l -> List.map (fun (_,f,_) -> f) l
   in
   let pkg_info global_options fields show_empty raw where
@@ -785,6 +782,17 @@ let show =
             | _ -> acc)
           [] atom_locs
       in
+      if opamfs = [] then
+        let dirnames =
+          OpamStd.List.filter_map (function
+              | `Dirname d -> Some (OpamFilename.Dir.to_string d)
+              | _ -> None)
+            atom_locs
+        in
+        OpamConsole.error_and_exit `Not_found "No opam files found at %s"
+          (OpamStd.List.concat_map ", " ~last_sep:" and "
+             (fun x -> x) dirnames )
+      else
       let errors, opams =
         List.fold_left (fun (errors,opams) opamf ->
             try

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -767,8 +767,11 @@ let show =
         OpamAuxCommands.simulate_autopin ~quiet:no_lint ~for_view:true st
           atom_locs
       in
-      OpamListCommand.info st
-        ~fields ~raw ~where ~normalise ~show_empty ~all_versions atoms;
+      if atoms = [] then
+        OpamConsole.error_and_exit `Not_found "No package found"
+      else
+        OpamListCommand.info st
+          ~fields ~raw ~where ~normalise ~show_empty ~all_versions atoms;
       `Ok ()
     | atom_locs, true ->
       if List.exists (function `Atom _ -> true | _ -> false) atom_locs then

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -751,8 +751,10 @@ let show =
          let opam = OpamFile.OPAM.read_from_channel stdin in
          print_just_file None opam;
          `Ok ()
-       with | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _  ->
-         OpamConsole.error_and_exit `File_error "Stdin parsing failed")
+       with
+       | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _ as exn ->
+         OpamConsole.error_and_exit `File_error
+           "Stdin parsing failed:\n%s" (Printexc.to_string exn))
     | atom_locs, false ->
       let fields, show_empty =
         if list_files then
@@ -801,19 +803,22 @@ let show =
             try
               errors, (Some opamf, (OpamFile.OPAM.read opamf))::opams
             with
-            | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _  ->
-              opamf::errors, opams)
+            | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _ as exn ->
+              (opamf, exn)::errors, opams)
           ([],[]) opamfs
       in
       List.iter (fun (f,o) -> print_just_file f o) opams;
       (if errors <> [] then
          let sgl = match errors with [_] -> true | _ -> false in
-         let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
+         let tostr (opamf, error) =
+           (OpamFilename.to_string (OpamFile.filename opamf))
+           ^ ":\n" ^ Printexc.to_string error
+         in
          OpamConsole.error "Parsing error on%s:%s"
            (if sgl then "" else "some opam files")
-           (match files with
-            | [f] -> " " ^ f
-            | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs));
+           (match errors with
+            | [f] -> tostr f
+            | fs -> OpamStd.Format.itemize tostr fs));
       if opams = [] then
         OpamStd.Sys.exit_because `File_error
       else

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -795,17 +795,16 @@ let show =
           ([],[]) opamfs
       in
       List.iter (fun (f,o) -> print_just_file f o) opams;
-      if errors = [] then
-        `Ok ()
-      else
-      let sgl = List.length errors = 1 in
-      let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
-      OpamConsole.error_and_exit `File_error "%s file%s failed to parse:%s"
-        (if sgl then "A" else "Some")
-        (if sgl then "" else "s")
-        (match files with
-         | [f] -> " "^ f
-         | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs)
+      (if errors <> [] then
+        let sgl = List.length errors = 1 in
+        let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
+        OpamConsole.error "%s file%s failed to parse:%s"
+          (if sgl then "A" else "Some")
+          (if sgl then "" else "s")
+          (match files with
+           | [f] -> " "^ f
+           | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs));
+      `Ok ()
   in
   Term.(ret
           (const pkg_info $global_options $fields $show_empty $raw $where

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -804,15 +804,17 @@ let show =
       in
       List.iter (fun (f,o) -> print_just_file f o) opams;
       (if errors <> [] then
-        let sgl = List.length errors = 1 in
-        let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
-        OpamConsole.error "%s file%s failed to parse:%s"
-          (if sgl then "A" else "Some")
-          (if sgl then "" else "s")
-          (match files with
-           | [f] -> " "^ f
-           | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs));
-      `Ok ()
+         let sgl = match errors with [_] -> true | _ -> false in
+         let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
+         OpamConsole.error "Parsing error on%s:%s"
+           (if sgl then "" else "some opam files")
+           (match files with
+            | [f] -> " " ^ f
+            | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs));
+      if opams = [] then
+        OpamStd.Sys.exit_because `File_error
+      else
+        `Ok ()
   in
   Term.(ret
           (const pkg_info $global_options $fields $show_empty $raw $where

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -707,15 +707,11 @@ let show =
       OpamConsole.warning "No opam files found in %s"
         (OpamFilename.Dir.to_string d);
       []
-    | l -> List.map (fun (_,f,_) -> Some f) l
+    | l -> List.map (fun (_,f,_) -> f) l
   in
   let pkg_info global_options fields show_empty raw where
       list_files file normalise no_lint just_file all_versions sort atom_locs =
-    let print_just_file f =
-      let opam = match f with
-        | Some f -> OpamFile.OPAM.read f
-        | None -> OpamFile.OPAM.read_from_channel stdin
-      in
+    let print_just_file opamf opam =
       if not no_lint then OpamFile.OPAM.print_errors opam;
       let opam =
         if not sort then opam else
@@ -723,9 +719,9 @@ let show =
       in
       if where then
         OpamConsole.msg "%s\n"
-          (match f with
-           | Some f ->
-             OpamFilename.(Dir.to_string (dirname (OpamFile.filename f)))
+          (match opamf with
+           | Some opamf ->
+             OpamFilename.(Dir.to_string (dirname (OpamFile.filename opamf)))
            | None -> ".")
       else
       let opam_content_list = OpamFile.OPAM.to_list opam in
@@ -754,8 +750,12 @@ let show =
     | [], false ->
       `Error (true, "required argument PACKAGES is missing")
     | [], true ->
-      print_just_file None;
-      `Ok ()
+      (try
+         let opam = OpamFile.OPAM.read_from_channel stdin in
+         print_just_file None opam;
+         `Ok ()
+       with | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _  ->
+         OpamConsole.error_and_exit `File_error "Stdin parsing failed")
     | atom_locs, false ->
       let fields, show_empty =
         if list_files then
@@ -777,16 +777,35 @@ let show =
       if List.exists (function `Atom _ -> true | _ -> false) atom_locs then
         `Error (true, "packages can't be specified with --just-file")
       else
-      let opams =
+      let opamfs =
         List.fold_left (fun acc al ->
             match al with
-            | `Filename f -> Some (OpamFile.make f) :: acc
+            | `Filename f -> (OpamFile.make f) :: acc
             | `Dirname d -> opam_files_in_dir d @ acc
             | _ -> acc)
           [] atom_locs
       in
-      List.iter print_just_file opams;
-      `Ok ()
+      let errors, opams =
+        List.fold_left (fun (errors,opams) opamf ->
+            try
+              errors, (Some opamf, (OpamFile.OPAM.read opamf))::opams
+            with
+            | Parsing.Parse_error | OpamLexer.Error _ | OpamPp.Bad_format _  ->
+              opamf::errors, opams)
+          ([],[]) opamfs
+      in
+      List.iter (fun (f,o) -> print_just_file f o) opams;
+      if errors = [] then
+        `Ok ()
+      else
+      let sgl = List.length errors = 1 in
+      let files = List.map (OpamFile.filename @> OpamFilename.to_string) errors in
+      OpamConsole.error_and_exit `File_error "%s file%s failed to parse:%s"
+        (if sgl then "A" else "Some")
+        (if sgl then "" else "s")
+        (match files with
+         | [f] -> " "^ f
+         | fs -> "\n"^OpamStd.Format.itemize (fun x -> x) fs)
   in
   Term.(ret
           (const pkg_info $global_options $fields $show_empty $raw $where

--- a/src/client/opamPinCommand.ml
+++ b/src/client/opamPinCommand.ml
@@ -32,9 +32,11 @@ let read_opam_file_for_pinning ?(quiet=false) name f url =
     let dir = OpamFilename.dirname (OpamFile.filename f) in
     (* don't add aux files for [project/opam] *)
     let add_files = OpamUrl.local_dir url = Some dir in
-    OpamStd.Option.map
+    let opam =
       (OpamFormatUpgrade.opam_file_with_aux ~quiet ~dir ~files:add_files
-         ~filename:f) (OpamFile.OPAM.read_opt f)
+         ~filename:f) (OpamFile.OPAM.safe_read f)
+    in
+    if opam = OpamFile.OPAM.empty then None else Some opam
   in
   (match opam0 with
    | None ->

